### PR TITLE
AAE-23521 Fix improve dropdown reactive form

### PR DESCRIPTION
**Pull request by kathrine0**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [x] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [x] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [x] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)

AAE-23521

**What is the new behaviour?**

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by eromano** [comment]
does the https://github.com/Alfresco/alfresco-ng2-components/tree/develop/lib/process-services/src/lib/form/widgets/dropdown needs any change with this modify?

**Review by mauriziovitale — APPROVED** [review]

File: lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.ts [hunk 1]
```diff
@@ -72,29 +87,145 @@ export class DropdownCloudWidgetComponent extends WidgetComponent implements OnI
...
+    private initFormControl() {
+        if (this.field?.required) {
+            this.dropdownControl.addValidators([Validators.required]);
+        }
+
+        if (this.field?.readOnly || this.readOnly) {
+            this.dropdownControl.disable({ emitEvent: false });
+        }
+
+        this.dropdownControl.valueChanges
```
Comment (tomgny) [inline lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.ts:172]: maybe we can move this subscriptions to separate method? what do you think?

File: lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.html [hunk 2]
```diff
@@ -1,43 +1,50 @@
...
+            <error-widget
+                class="adf-dropdown-failed-message"
+                *ngIf="isRestApiFailed"
+                required="{{ 'FORM.FIELD.REST_API_FAILED' | translate: { hostname: restApiHostName } }}"
+            ></error-widget>
+            <error-widget
+                class="adf-dropdown-failed-message"
+                *ngIf="variableOptionsFailed"
+                required="{{ 'FORM.FIELD.VARIABLE_DROPDOWN_OPTIONS_FAILED' | translate }}"
+            ></error-widget>
```
Comment (tomgny) [inline lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.html:47]: i think we can display errors using only one error-widget (by assign errors to the validationSummary property), but its something optional, in the future i believe we will replace this by mat-error ([we have ticket for it](https://hyland.atlassian.net/browse/AAE-24762))
so its just a suggestion

File: lib/core/src/lib/form/components/widgets/core/form-field-validator.ts [hunk 3]
```diff
@@ -51,22 +50,6 @@ export class RequiredFieldValidator implements FormFieldValidator {
 
     validate(field: FormFieldModel): boolean {
         if (this.isSupported(field) && field.isVisible) {
-            if (field.type === FormFieldTypes.DROPDOWN) {
-                if (field.hasMultipleValues) {
```
Comment (tomgny) [inline lib/core/src/lib/form/components/widgets/core/form-field-validator.ts:55]: just a note, please make sure we still have a correct behaviour of form validity as a whole, like disable 'Complete' button where `field.hasMultipleValues` and `Array.isArray(field.value) && !!field.value.length` return false

File: lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.ts [hunk 4]
```diff
@@ -72,29 +87,145 @@ export class DropdownCloudWidgetComponent extends WidgetComponent implements OnI
...
+            this.field.validationSummary = new ErrorMessageModel('');
+            return;
+        }
+
+        if (this.dropdownControl.invalid && this.dropdownControl.errors.required) {
+            this.field.validationSummary = new ErrorMessageModel({ message: 'FORM.FIELD.REQUIRED' });
+        }
+    }
+
+    private initFilter() {
```
Comment (tomgny) [inline lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.ts:204]:
```suggestion
    private initFilter(): void {
```

File: lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.ts [hunk 5]
```diff
@@ -72,29 +87,145 @@ export class DropdownCloudWidgetComponent extends WidgetComponent implements OnI
...
+                filter(() => !!this.field),
+                takeUntil(this.onDestroy$)
+            )
+            .subscribe(() => this.handleErrors());
+
+        this.dropdownControl.setValue(this.field?.value, { emitEvent: false });
+        this.handleErrors();
+    }
+
+    private handleErrors() {
```
Comment (tomgny) [inline lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.ts:193]:
```suggestion
    private handleErrors(): void {
```

**Review by tomgny — COMMENTED** [review]
added some notes and small suggestions

**Comment by kathrine0** [comment]
> does the https://github.com/Alfresco/alfresco-ng2-components/tree/develop/lib/process-services/src/lib/form/widgets/dropdown needs any change with this modify?

That's right, I modified the dropdown widget as well

File: lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.html [hunk 6]
```diff
@@ -1,43 +1,50 @@
-<div class="adf-dropdown-widget {{field.className}}"
-     [class.adf-invalid]="(!field.isValid && isTouched()) || isRestApiFailed" [class.adf-readonly]="field.readOnly" [class.adf-left-label-input-container]="field.leftLabels">
-    <div class="adf-dropdown-widget-top-labels">
-        <label class="adf-label" [attr.for]="field.id" [class.adf-left-label]="field.leftLabels">{{field.name | translate }}<span class="adf-asterisk"
-            *ngIf="isRequired()">*</span>
+<div
+    class="adf-dropdown-widget {{field.className}}"
+    [class.adf-invalid]="dropdownControl.invalid && dropdownControl.touched"
```
[the remark this answers is not in the document]
Comment (kathrine0) [reply lib/process-services-cloud/src/lib/form/components/widgets/dropdown/dropdown-cloud.widget.html:3]: I think it's not needed here - no additional code

File: lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts [hunk 7]
```diff
@@ -85,29 +105,77 @@ export class DropdownWidgetComponent extends WidgetComponent implements OnInit {
...
-            optionValue = option.name;
-        }
-        return optionValue;
+    private isReadOnlyForm(): boolean {
+        return !!this.field?.form?.readOnly;
     }
 
-    isReadOnlyType(): boolean {
-        return this.field.type === 'readonly';
+    private initFormControl() {
```
Comment (tomgny) [inline lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts:112]:
```suggestion
    private initFormControl(): void {
```

File: lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts [hunk 8]
```diff
@@ -85,29 +105,77 @@ export class DropdownWidgetComponent extends WidgetComponent implements OnInit {
...
+            )
+            .subscribe(() => this.handleErrors());
+
+        this.dropdownControl.setValue(this.getOptionValue(this.field?.value), { emitEvent: false });
+        this.handleErrors();
     }
 
-    showRequiredMessage(): boolean {
-        return (this.isInvalidFieldRequired() || this.field.value === 'empty') && this.isTouched();
+    private handleErrors() {
```
Comment (tomgny) [inline lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts:142]:
```suggestion
    private handleErrors(): void {
```

File: lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts [hunk 9]
```diff
@@ -85,29 +105,77 @@ export class DropdownWidgetComponent extends WidgetComponent implements OnInit {
...
+        }
+
+        if (this.dropdownControl.invalid && this.dropdownControl.errors.required) {
+            this.field.validationSummary = new ErrorMessageModel({ message: 'FORM.FIELD.REQUIRED' });
+        }
     }
 
-    get isReadOnlyField(): boolean {
-        return this.field.readOnly;
+    private setOptionValue(option: string | FormFieldOption, field: FormFieldModel) {
```
Comment (tomgny) [inline lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts:157]:
```suggestion
    private setOptionValue(option: string | FormFieldOption, field: FormFieldModel): void {
```

File: lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts [hunk 10]
```diff
@@ -85,29 +105,77 @@ export class DropdownWidgetComponent extends WidgetComponent implements OnInit {
...
+            field.value = option.id;
+            return;
+        }
+
+        field.value = option.name;
     }
 
-    private isReadOnlyForm(): boolean {
-        return !!this.field?.form?.readOnly;
+    private getOptionValue(value?: string | FormFieldOption) {
```
Comment (tomgny) [inline lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.ts:170]:
```suggestion
    private getOptionValue(value?: string | FormFieldOption): FormFieldOption | undefined {
```

File: lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.html [hunk 11]
```diff
@@ -1,23 +1,16 @@
...
-                    [disabled]="field.readOnly"
-                    (ngModelChange)="onFieldChanged(field)"
-                    (blur)="markAsTouched()">
-            <mat-option *ngFor="let opt of field.options"
-                        [value]="getOptionValue(opt, field.value)"
-                        [id]="opt.id">{{opt.name}}
-            </mat-option>
-            <mat-option id="readonlyOption" *ngIf="isReadOnlyType()" [value]="field.value">{{field.value}}</mat-option>
+        <mat-select class="adf-select" [id]="field.id" [formControl]="dropdownControl">
+            <mat-option *ngFor="let opt of field.options" [value]="opt" [id]="opt.id">{{opt.name}} </mat-option>
```
Comment (tomgny) [inline lib/process-services/src/lib/form/widgets/dropdown/dropdown.widget.html:9]:
```suggestion
            <mat-option *ngFor="let opt of field.options" [value]="opt" [id]="opt.id">{{opt.name}}</mat-option>
```
:D

**Review by tomgny — APPROVED** [review]
good job

File: lib/core/src/lib/form/components/widgets/error/error.component.scss [hunk 12]
```diff
@@ -1,5 +1,6 @@
 .adf-error {
     display: flex;
+    align-items: center;
```
Comment (DenysVuika) [inline lib/core/src/lib/form/components/widgets/error/error.component.scss:3]: is this expected? the error under each field should be centered, not left-aligned?

**Pull request merged**